src/aws_signature_utils.erl

```diff
@@ -52,18 +52,24 @@ hex(N, upper) when N < 16 ->
 %%
 %% An alternative to `uri_string:parse/1' to support OTP below 21.
 -spec parse_url(binary()) -> #{host => binary(), path => binary(), query => binary()}.
-parse_url(URL) when is_binary(URL) ->
-    %% From https://datatracker.ietf.org/doc/html/rfc3986#appendix-B
-    {ok, Regex} = re:compile("^(([a-z][a-z0-9\\+\\-\\.]*):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?", [caseless]),
-
-    case re:run(URL, Regex, [{capture, all, binary}]) of
-        {match, [_, _1, _2, _3, Authority, Path, _6, Query | _]} ->
-            #{host => Authority, path => Path, query => Query};
-        {match, [_, _1, _2, _3, Authority, Path | _]} ->
-            #{host => Authority, path => Path, query => <<"">>};
-        _ ->
-            #{host => <<"">>, path => <<"">>, query => <<"">>}
-    end.
+-ifdef(OTP_RELEASE). % OTP >= 21
+  parse_url(URL) when is_binary(URL) ->
+    #{host := Host, path := Path} = P = uri_string:parse(URL),
+    #{host => Host, path => Path, query => maps:get(query, P, <<>>)}.
+-else. % OTP < 21
+  parse_url(URL) when is_binary(URL) ->
+    {ok, {_Scheme, _UserInfo, Host, _Port, Path0, Query0}} = http_uri:parse(binary_to_list(URL)),
```

Comment:
When the URL has trailing slash as in `http://a.b/` the path is `/`, otherwise as in `http://a.b` the path is empty. Both `uri_string:parse` and the regexp handle it correctly, however `http_uri:parse` doesn't distinguish those cases.

Also, I think patching the parsing result like this is brittle, so I'd suggest deferring to the current implementation on OTP < 21, so that we use `uri_string:parse` as an optimisation on more recent versions. wdyt?

Reply:
Totally agree it's brittle and not my brightest work. I fix :+1: